parsers/moose.c

```diff
@@ -0,0 +1,273 @@
+		moose->packageCork = corkIndex;
+	else if ((!moose->notInMoose) && tag->kindIndex == KIND_PERL_SUBROUTINE)
+	{
+		tagEntryInfo moose_e;
+		initTagEntry (&moose_e, tag->name, K_METHOD);
+		moose_e.extensionFields.scopeIndex = moose->classCork;
+		makeTagEntry (&moose_e);
+	}
+}
+
+static void enteringPodNotify (perlSubparser *perl)
+{
+	struct mooseSubparser *moose = (struct mooseSubparser *)perl;
+	moose->inPod = true;
+}
+
+static void leavingPodNotify  (perlSubparser *perl)
+{
+	struct mooseSubparser *moose = (struct mooseSubparser *)perl;
+	moose->inPod = false;
+}
+
+static bool leaveMoose (const char *line CTAGS_ATTR_UNUSED,
+						const regexMatch *matches CTAGS_ATTR_UNUSED,
+						unsigned int count CTAGS_ATTR_UNUSED,
+						void *data)
+{
+	struct mooseSubparser *moose = data;
+
+	if (moose->inPod)
+		return true;
+
+	tagEntryInfo *e = getEntryInCorkQueue (moose->classCork);
+	Assert (e);
+	e->extensionFields.endLine = getInputLineNumber ();
+
+	moose->classCork = CORK_NIL;
+	moose->notInMoose = true;
+	moose->packageCork = CORK_NIL;
+	return true;
+}
+
+static bool enterMoose(const char *line CTAGS_ATTR_UNUSED,
+						const regexMatch *matches CTAGS_ATTR_UNUSED,
+						unsigned int count CTAGS_ATTR_UNUSED,
+						void *data)
+{
+	struct mooseSubparser *moose = data;
+
+	if (moose->inPod)
+		return true;
+
+	if (moose->packageCork == CORK_NIL)
+		return true;
+
+	moose->notInMoose = false;
+	tagEntryInfo *perl_e = getEntryInCorkQueue (moose->packageCork);
+	Assert (perl_e);
+
+	tagEntryInfo moose_e;
+	initTagEntry (&moose_e, perl_e->name, K_CLASS);
+	moose_e.lineNumber = perl_e->lineNumber;
+	moose_e.filePosition = perl_e->filePosition;
+	moose->classCork = makeTagEntry (&moose_e);
+
+	return true;
+}
+
+static bool extendClass(const char *line,
+						const regexMatch *matches,
+						unsigned int count,
+						void *data)
+{
+	struct mooseSubparser *moose = data;
+
+	if (moose->inPod)
+		return true;
+
+	if (moose->classCork == CORK_NIL)
+		return true;
+
+	tagEntryInfo *e = getEntryInCorkQueue (moose->classCork);
+	Assert (e);
+	if (e->extensionFields.inheritance == NULL)
+		e->extensionFields.inheritance = eStrndup (line + matches[1].start,
+												   matches[1].length);
+
+	return true;
+}
+
+static bool findAttribute(const char *line,
+						  const regexMatch *matches,
+						  unsigned int count,
+						  void *data)
+{
+	struct mooseSubparser *moose = data;
+
+	if (moose->inPod)
+		return true;
+
+	if (count < 3)
+		return true;
+
+	tagEntryInfo e;
+	char *buf = eStrndup (line + matches[2].start, matches[2].length);
+
+	initTagEntry (&e, buf, K_ATTR);
+	if (moose->classCork != CORK_NIL)
+		e.extensionFields.scopeIndex = moose->classCork;
+	makeTagEntry (&e);
+
+	eFree (buf);
+
+	return true;
+}
+
+static void findMooseTags(void)
+{
+	scheduleRunningBaseparser (RUN_DEFAULT_SUBPARSERS);
+}
+
+static void initializeMooseParser (langType language)
+{
+	addLanguageCallbackRegex (language, "^ *use +Moose(:| *;)",
```

Comment:
I'm not sure how this syntax works, but this sublanguage which we've been referring to in the ticket as "Moose et al" should also trigger on `Moo` which also provides all these same features. https://metacpan.org/pod/Moo#IMPORTED-SUBROUTINES

Reply:
Maybe just "use Moose;" is enough.  I should remove "Moose:".

I don't know Moo but I guess we should implement is as a different subparser running on Perl.
even if Moo and Moose subparesrs can share C functions and data types definitions.

Reply:
"Moose:" may have a sense when we support Rules.

Reply:
Moo and Mouse essentially follow the same spec as Moose. So much so that I think they both naturally use Moose if it's there. (it becomes a noop) The primary reason for their existence is to shift work into compile time and eliminate work from run-time, but they coat-tail (follow very closely) the Moose syntax.

Reply:
@masatake we're in a situation where I either have to duplicate the patch for a new subparser "Moo" that works that the same on the initial push. Or, we can add Moo to the trigger so it works on both. The difference between the two is pretty minimal. I don't see the purpose of maintaining two subparsers, but we are losing a lot of users of this because it's not yet catching Moo and Mouse. Please advise. If you want me to duplicate the parser (and thus have one subparser for each Moose we'll quickly stack up a few highly duplicate code bases).

Reply:
New issue for support here https://github.com/universal-ctags/ctags/issues/2103

Reply:
Do you mean extending following regex patterns is enough to support Moo and Mouse?
```
	addLanguageCallbackRegex (language, "^[ \t]*use +Moose *;",
							  "{exclusive}",
							  enterMoose, NULL,
							  &mooseSubparser);
	addLanguageCallbackRegex (language, "^[ \t]*no +Moose *;",
							  "{exclusive}",
							  leaveMoose, &mooseSubparser.notInMoose,
							  &mooseSubparser);
```
?

Reply:
I think gathering common code to somewhere (let's call it perlMx) is better.
We can derive Moose, Mouse, and Moo from perlMx.
I understand Moose, Mouse and Moo has very similar syntax.

The question is the biggest difference betwwen them.
If really no diffrence between them, we should not take much time to make extra subparsers.
If there is, separating multipe subparsers has make sense to make potential cost of maintenance in the future minimal.

Reply:
I found an interesting page:
https://blog.jdpfu.com/2011/12/02/of-moose-mouse-moo-and-mo-for-oo-perl

They are not the same.

Reply:
As far as reading the page, I think we should build a small library for implementing Moose, Moo, Mouse and Mo(?).

We can write each parsers using the library.
Deciding the name for the library is the start point.

Reply:
Having multiple object systems in a language. It looks a bit crazy situation. But I just remember "There's More Than One Way To Do It."
This is the perl way.

Reply:
@masatake  they are not the same because of their implementation. The API into them is the same. For example, there would be no difference in the patch you've provided to support Moose and Moo other than the enter language line. In fact, in my use case here I don't even care about `Moose`. I've verified your code works by changing that Moose to `(Moose|Moo)`.

Reply:
> I've verified your code works by changing that Moose to (Moose|Moo).

How about Mouse?

Reply:
I think we can and **should** add Mouse to it. None of these libraries have any significant change in API. The differences are **extremely** minor, and all of these people benefit from your code.

Really 99% of the functional advantage of this patch is simply in the attribute tag with `has`. That's cool. It'll work on any Moose-like module build system. The remaining useful functionality is mostly the around/before/after method modifiers.

Mouse hasn't been patched since August of last year. Mouse exists so people can write modern perl objects into actual 1992 CGI scripts. It's not a very high visibility project and Perl as a whole is on the decline. I would enable the subparser for Mouse/Moo/Moose and it'll be good enough.

Reply:
I see. Just extending the regex pattern may be enough. However, as a field value, `language:Moose` is reported for all 3 object systems. "Different things should be tagged differently", is the motto of u-ctags.
Reporting the language field differently for 3 object systems like "language:Moose", "language"Moo", and "language:Moo",  will be better in this case
See also #2079 and 2076.

How about the patch at the end of this page?